### PR TITLE
Create separate record for each notification

### PR DESCRIPTION
**Pull request by smirnovaka**
https://github.com/nanoPayinc/NANOPAY/issues/2264
@kgrgreer

### CODE REVIEW

**Comment by smirnovaka** [comment]
https://github.com/nanoPayinc/NANOPAY/issues/2250

File: src/foam/nanos/notification/SendNotificationDAO.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package foam.nanos.notification;
+
+import foam.core.Detachable;
+import foam.core.FObject;
+import foam.core.X;
+import foam.dao.AbstractSink;
+import foam.dao.DAO;
+import foam.dao.ProxyDAO;
+import foam.nanos.auth.User;
+
```
Comment (kgrgreer) [inline src/foam/nanos/notification/SendNotificationDAO.java]: delete empty lines between imports

File: src/foam/nanos/notification/SendNotificationDAO.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import static foam.mlang.MLang.EQ;
+
+public class SendNotificationDAO extends ProxyDAO {
+
+
+  public SendNotificationDAO(DAO delegate) {
```
Comment (kgrgreer) [inline src/foam/nanos/notification/SendNotificationDAO.java]: Do we need this constructor?

File: src/foam/nanos/notification/SendNotificationDAO.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    if ( notif.getBroadcasted() ) {
+      userDAO.select(new AbstractSink() {
+        @Override
+        public void put(Object o, Detachable d) {
+          User user = (User) o;
+          send(user, notif, x);
+        }
+      });
+    } else if ( notif.getGroupId() != null ){
```
Comment (kgrgreer) [inline src/foam/nanos/notification/SendNotificationDAO.java]: space between ) and {

File: src/foam/nanos/notification/SendNotificationDAO.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        @Override
+        public void put(Object o, Detachable d) {
+          User user = (User) o;
+          send(user, notif, x);
+        }
+      });
+    }
+
+    if ( notif.getGroupId() == null && ! notif.getBroadcasted() ) {
+      return super.put_(x,notif);
```
Comment (kgrgreer) [inline src/foam/nanos/notification/SendNotificationDAO.java]: space after ,

File: src/foam/nanos/notification/SendNotificationDAO.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    return obj;
+  }
+
+  public void send(User user, Notification notif, X x) {
+    Notification notification = new Notification();
+    notification.setUserId(user.getId());
```
Comment (kgrgreer) [inline src/foam/nanos/notification/SendNotificationDAO.java]: Would this be shorter and less suseptable to breakage when new fields are added if you used copyFrom() to copy the values, or just did an fclone() of notif() and then made changes to the recipient fields?

File: src/services [hunk 6]
```diff
@@ -74,5 +74,5 @@ p({"class":"foam.nanos.boot.NSpec", "name":"verifyEmail",                      "
...
-p({"class":"foam.nanos.boot.NSpec", "name":"notificationDAO",                   "lazy":true,  "serve":true,  "serviceScript":"return new foam.dao.PMDAO(x, new foam.dao.SequenceNumberDAO (new foam.nanos.notification.SendEmailNotificationDAO(new foam.dao.JDAO(x, foam.nanos.notification.Notification.getOwnClassInfo(), \"notifications\"))));", "client":"{\"of\":\"foam.nanos.notification.Notification\"}"})
+p({"class":"foam.nanos.boot.NSpec", "name":"notificationDAO",                   "lazy":true,  "serve":true,  "serviceScript":"return new foam.dao.PMDAO(x, new foam.dao.SequenceNumberDAO(new foam.nanos.notification.SendNotificationDAO(new foam.nanos.notification.SendEmailNotificationDAO(new foam.dao.JDAO(x, foam.nanos.notification.Notification.getOwnClassInfo(), \"notifications\")))));", "client":"{\"of\":\"foam.nanos.notification.Notification\"}"})
```
Comment (kgrgreer) [inline src/services]: In the event that it's a group or broadcast, a sequence number will still be claimed but unused. Putting the SequenceNumberDAO below the SendNotificationDAO would avoid this problem.

**Pull request merged**